DataDog/datadog-agent
# [AMLII-1994] Improve expected_tags_duration for metrics

### What does this PR do?
Fixes configurable host tags that are added to metrics expected_tags_duration is configured.

This PR changes the logic for adding host tags and moves it after aggregation occurs so that a metric's context hash does not change and the aggregator does not mistake the same metric as a new one. 

### Motivation
[Issue](https://datadoghq.atlassian.net/jira/software/c/projects/AMLII/boards/5315?selectedIssue=AMLII-1994)
### Describe how to test/QA your changes

**To verify that configurable host tags work as expected**

1. Add the following settings to datadog.yaml
```
tags: "foo:bar"
expected_tags_duration: 1m
log_level: 'debug'
log_payloads: true 
```
2. Start the agent, and tail the logs.
3. grep for tags `foo:bar`
4. For the first ~1m of the agent running you should see foo:bar tags on every metric logged in the payloads.
5. After ~1m passes, you should no longer see these tags.

**To verify that the bug was resolved**
1. Use the same datadog.yaml settings as before
2. Create a test file
`dev/dist/checks.d/testcheck.py
`
3. Inside the file, configure the following check 
```
from datadog_checks.base import AgentCheck

class ServerCheck(AgentCheck):
    inc = 0  

    def check(self, instance):
        ServerCheck.inc += 100  
        self.monotonic_count('brian.test.check.count', ServerCheck.inc, flush_first_value=True)
```
4. Run the agent then the check
5. Verify that a spike does not occur in `1m`, here is how the spike used to look
<img width="701" alt="image" src="https://github.com/user-attachments/assets/4e5cf9f3-651d-44f1-aee3-85ac590b0ab9">



QA steps are also given in the [ticket](https://datadoghq.atlassian.net/jira/software/c/projects/AMLII/boards/5315?selectedIssue=AMLII-1994)

### Additional Notes

I've benchmarked these changes with dogstatsd stress tests and displayed the results in this [document](https://datadoghq.atlassian.net/wiki/spaces/~71202058eaddc626774b6382b0a98b3e9f4a84/pages/4221796722/Bench+Mark+Tests+for+enhancing+configurable+host+tags+new+version). 

[Notebook](https://dddev.datadoghq.com/notebook/10175380/andrew-oct-15-2024-15-13)

**RSS pod memory usage**

Old host tag feature:
- Tags enabled: 237M average
- Tags disabled: 228M average

New host tag feature:
- Tags enabled: 229M average
- Tags disabled 224M average

**CPU usage**

Old host tag feature
- Tags enabled: 1.18 cores average
- Tags disabled: 1.15 cores average

New host tag feature
- Tags enabled: 1.16 cores average
- Tags disabled: 1.11 cores average

Results:
3.7 % reduction in memory usage when tags are enabled while using the new host tag feature
1.8 % reduction in memory usage when tags are disabled while using the new host tag feature

1.7 % reduction in average cores used when tags are enabled while using the new host tag feature
3.5 % reduction in average cores used when tags are disabled while using the new host tag feature

The old version also has a [CPU drop](https://dddev.datadoghq.com/dashboard/tgn-spj-6dv/dogstatsd-agent-benchmarks?fromUser=false&fullscreen_end_ts=1729008099026&fullscreen_paused=true&fullscreen_refresh_mode=paused&fullscreen_section=overview&fullscreen_start_ts=1729005671121&fullscreen_widget=5269084395850462&refresh_mode=paused&tpl_var_scenario%5B0%5D=andrew%2A&from_ts=1729005671121&to_ts=1729008099026&live=false) when the host tags expire which the new version does not have.
